### DISCUSSION TITLE
enable_cell_picking and orig_extract_id problem

### DISCUSSION THREAD

**Discussion by panicbuttonvfx** [pyvista/pyvista#4883] [Q&A]
Hi, 

When using `enable_cell_picking`, I've noticed a new array is created name `orig_extract_id`. This array seems to contain the indices of the cells of the mesh, rather than the ones that are selected (comparing that array of indices with the one given from the entire mesh, they both match). Is that the intended behavior? If it is, is there a way to catch the indices of the cells being selected instead?  I'm trying to delete selected cells (triangles in this case) from a mesh.

I've found a partial way to do it with this piece of code, extracted from pyvista, but it only catches the cells that are facing to camera:

```
            x0, y0, x1, y1 = renderer.get_pick_position()
            selector = vtk.vtkOpenGLHardwareSelector()
            selector.SetFieldAssociation(vtk.vtkDataObject.FIELD_ASSOCIATION_CELLS)
            selector.SetRenderer(renderer)
            selector.SetArea(x0, y0, x1, y1)
            selection = selector.Select()
            for node in range(selection.GetNumberOfNodes()):
                self.cellids = selection.GetNode(node)
```

Thanks!

**Comment by banesullivan** [comment]
You're looking at the input mesh, not the produced mesh from the picking. Per the docs of `enable_cell_picking`:

> ... Selection will be saved to ``self.picked_cells`` ...


We add the `orig_extract_id` array to the input mesh then run it through the picking routine so that the output mesh will track each cells original IDs

-----

Here is an example of how to access the selection:

```py
import pyvista as pv
import numpy as np

mesh = pv.Sphere()

p = pv.Plotter(notebook=False)
p.add_mesh(mesh)
p.enable_cell_picking(through=True)
p.show()
```

![2020-06-01 19 04 33](https://user-images.githubusercontent.com/22067021/83463033-db264c00-a43a-11ea-8294-9a2a6ae7f81d.gif)


```py
picked = p.picked_cells
picked.plot(color=True, show_edges=True)
```
![download](https://user-images.githubusercontent.com/22067021/83463044-e4afb400-a43a-11ea-9f84-4bec3d5473d3.png)


Now you can use `orig_extract_id` to do things with the original mesh:

```py
mesh["foo"] = np.random.rand(mesh.n_cells)
mesh["foo"][picked["orig_extract_id"]] = 0.0
mesh.plot(scalars="foo")
```

![download](https://user-images.githubusercontent.com/22067021/83463133-1d4f8d80-a43b-11ea-95a7-239eafdc10b2.png)

**Comment by panicbuttonvfx** [comment]
Thanks Bane, that was it. I was simply getting the `orig_extract_id` from the original mesh instead of the picked one. Oh boy...

**Comment by abhishekmonogram** [comment]
How can I go about doing the same as what this post suggests but instead of using a the rectangular selection tool, draw a polygon in the surface and then extract the picked mesh? @banesullivan

**Comment by banesullivan** [comment]
@abhishekmonogram, please post a new discussion on the new Discussions page where the community is more active these days: https://github.com/pyvista/pyvista/discussions
